Remove SVGLengthContext's viewport for performance

Ahmad-S792 opened the pull request:
#### 5a9ffa94d048a37908ed8bbb6c0694b45feacda1
<pre>
Remove SVGLengthContext&apos;s viewport for performance

<a href="https://bugs.webkit.org/show_bug.cgi?id=279716">https://bugs.webkit.org/show_bug.cgi?id=279716</a>
<a href="https://rdar.apple.com/136009811">rdar://136009811</a>

Reviewed by Nikolas Zimmermann and Simon Fraser.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/cf8a467f2e89d9e9dd5924750b782dd3724935c0">https://chromium.googlesource.com/chromium/src.git/+/cf8a467f2e89d9e9dd5924750b782dd3724935c0</a>

This patch removes m_overriddenViewport from SVGLengthContext and
moves all code into the only user - SVGLengthContext::resolveRectangle.
This was possible because resolveRectangle was the only callsite to
use a viewport.

* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::resolveRectangle):
(WebCore::SVGLengthContext::convertValueToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromPercentageToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromPercentageToUserUnits):
(WebCore::SVGLengthContext::viewportSize const):
(WebCore::SVGLengthContext::computeViewportSize const):
* Source/WebCore/svg/SVGLengthContext.h:
* LayoutTests/platform/glib/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/platform/ios/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input.svg:

Canonical link: <a href="https://commits.webkit.org/283877@main">https://commits.webkit.org/283877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/565168c19394c0afd6d2e223c58e7ea9d4810077

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67148 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71183 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18281 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69266 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53844 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12301 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70215 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42792 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58100 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34379 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15494 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16635 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72886 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15193 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61317 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58159 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61394 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/15032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9130 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2730 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42332 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43409 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44595 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43150 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -211,7 +198,12 @@ ExceptionOr<float> SVGLengthContext::convertValueFromPercentageToUserUnits(float
+    return convertValueFromPercentageToUserUnits(value, lengthMode, *viewportSize);
+}
+
+float SVGLengthContext::convertValueFromPercentageToUserUnits(float value, SVGLengthMode lengthMode, FloatSize viewportSize)
```
This could just be a static function inside the file (rather than a static class method).

Ahmad-S792 replied:
Done in https://github.com/WebKit/WebKit/commit/0a6cceedab76e7444eab78de32fa00ed1deb05b7.

Ahmad-S792 replied:
@smfr - I tried to do this change and tried to see if I can do it in `.h` but it does not seems to compile. So I have undid this change as of right now. Any suggestions?

shallawa replied:
You need to define 
```
static float convertValueFromPercentageToUserUnits(...)
```
before it is called. Just move it before `SVGLengthContext::resolveRectangle()` around line 45.

Ahmad-S792 replied:
OK, let me retry this locally. :-)

Ahmad-S792 replied:
@shallawa - I tried but it does not compile:

<img width="661" alt="image" src="https://github.com/user-attachments/assets/07a3fc61-ece0-4aac-80af-c674bb770e9b">

and gives error:

<img width="1051" alt="image" src="https://github.com/user-attachments/assets/72870316-cb7e-4ea2-8358-92ca794c29f3">

shallawa replied:
I see. The error happens because there is another non static function with the same name at line 58 in the SVGContextLength.h
```
ExceptionOr<float> convertValueFromPercentageToUserUnits(float value, SVGLengthMode) const;
```
So the compiler assumes your calls are for the class method. But the class method takes two arguments and you are passing three. So it asks if you meant to call the static one `::WebCore::convertValueFromPercentageToUserUnits()`.

I guess you can leave this method as a static class method for now.

Ahmad-S792 replied:
Thanks!

smfr commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -41,27 +41,22 @@ SVGLengthContext::SVGLengthContext(const SVGElement* context)
-        SVGLengthContext lengthContext(context);
-        return FloatRect(x.value(lengthContext), y.value(lengthContext), width.value(lengthContext), height.value(lengthContext));
+    if (type != SVGUnitTypes::SVG_UNIT_TYPE_USERSPACEONUSE) {
+        const FloatSize& viewportSize = viewport.size();
```
`auto viewportSize = viewport.size();`

Ahmad-S792 replied:
Done in https://github.com/WebKit/WebKit/commit/0a6cceedab76e7444eab78de32fa00ed1deb05b7.

nikolaszimmermann commented on LayoutTests/platform/mac/svg/batik/text/textEffect3-expected.txt:
```diff
@@ -99,7 +99,7 @@ layer at (0,0) size 450x500
             chunk 1 (middle anchor) text run 1 at (150.74,340.00) startOffset 0 endOffset 1 width 35.57: "K"
         RenderSVGInlineText {#text} at (0,0) size 0x0
-          [filter="dropShadow"] RenderSVGResourceFilter {filter} at (-45,-50) size 540x600
+          [filter="dropShadow"] RenderSVGResourceFilter {filter} at (0,0) size 0x0
```
Where do these changes come from?

Ahmad-S792 replied:
I tried to see visually using here - https://ews-build.s3-us-west-2.amazonaws.com/macOS-Ventura-Release-WK2-Tests-EWS/ed87b7ae-15732-rerun/retries/svg/batik/text/textEffect3-diffs.html but it seems that it is just text readjustment without any difference in rendering, so I was OK with just rebaseline.

nikolaszimmermann commented on LayoutTests/svg/filters/feMerge-wrong-input.svg:
```diff
@@ -1,7 +1,7 @@
 <defs>
 <!-- check that WebKit doesn't crash, if an effect is not available in feMerge -->
-<filter id="filter" primitiveUnits="objectBoundingBox">
+<filter id="filter" filterUnits="userSpaceOnUse" primitiveUnits="objectBoundingBox">
```
Any particular reason why you have to modify this?

Ahmad-S792 replied:
So original Blink's patch added `viewport.isEmpty()` (which I didn't added in first place), because I knew it would lead to cause another issue, which they fixed in following PR:

https://source.chromium.org/chromium/chromium/src/+/8ff0050993b2d2654c9f3170e789e65b0b303691

So in the above fix, they have to adjust this test as well. So that's what I did.

Ahmad-S792 committed:
Remove SVGLengthContext's viewport for performance

https://bugs.webkit.org/show_bug.cgi?id=279716
rdar://136009811

Reviewed by Nikolas Zimmermann and Simon Fraser.

Merge: https://chromium.googlesource.com/chromium/src.git/+/cf8a467f2e89d9e9dd5924750b782dd3724935c0

This patch removes m_overriddenViewport from SVGLengthContext and
moves all code into the only user - SVGLengthContext::resolveRectangle.
This was possible because resolveRectangle was the only callsite to
use a viewport.

* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::resolveRectangle):
(WebCore::SVGLengthContext::convertValueToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromPercentageToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromPercentageToUserUnits):
(WebCore::SVGLengthContext::viewportSize const):
(WebCore::SVGLengthContext::computeViewportSize const):
* Source/WebCore/svg/SVGLengthContext.h:
* LayoutTests/platform/glib/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/platform/ios/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/platform/mac/svg/batik/text/textEffect3-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input.svg:

Canonical link: https://commits.webkit.org/283877@main
